Add middle attributes sorting

SkyaTura opened the pull request:
## TODO

- [ ] Explain this PR

#### Original post
<details>
## Reason for this PR

Every time this code is saved it gets a different result:

```vue
<template lang="pug">
Foo(c b a d i f g h e l n)
</template>
```

## Expected result

```vue
<template lang="pug">
Foo(a b c d e f g h i l n)
</template>
```
<details>

SkyaTura commented:
I've accidentally formatted things that I shouldn't on a formatter library :P

Going to fix it right now!

Shinigami92 commented:
<img src="https://user-images.githubusercontent.com/7195563/94648028-f7a39980-02f1-11eb-8624-7b680336044a.png" width="400px"/>

1. Please use draft PRs
2. Please request me as a reviewer if you are done

:slightly_smiling_face:

Shinigami92 commented:
Please add a prettier config to the issue descriptions

Currently I would expect that nothing would get sorted cause the default would be taken

Also please note that the sorting isn't sorting in an alphabetical order in any way

Shinigami92 commented on src/options/attribute-sorting/utils.ts:
```diff
@@ -22,47 +22,38 @@ function compareByIndex(leftIndex: number, rightIndex: number): CompareResult {
-				break;
-			}
-		}
+	if (beginning > 0) return -1;
```
Oh I need to improve my eslint...

I really don't like these non curly braces ^^'

SkyaTura replied:
I'll add those later ;)

Shinigami92 commented on src/options/attribute-sorting/utils.ts:
```diff
@@ -22,47 +22,38 @@ function compareByIndex(leftIndex: number, rightIndex: number): CompareResult {
-	arr.length = 0;
-	arr.push(...preSorted.concat(sorted).concat(postSorted));
-	return arr;
+	const preSort: T[] = [...arr].slice(0, start);
```
Please explain why these three `[...arr]` are used, I want to understand :slightly_smiling_face:

SkyaTura replied:
I reverted this change, it was useless.

Shinigami92 commented on tests/options/sortAttributes/sort-attributes.test.ts:
```diff
@@ -38,19 +38,62 @@ describe('Options', () => {
+				'z',
+				'a',
+				'b',
+				'c',
```
Why would `a` sorted before `c` if you didn't configured it in any way in `beginning` or `end`?
I won't expect this `expected`

I would expect this: `'x', 'y', 'z', 'c', 'a', 'b', 'v-for', ':key', 'src', 'alt', '@click', ':disabled'`

SkyaTura replied:
I was considering that middle attributes should be sorted alphabetically, which I was, indeed, wrong.

_this is what happens when you try to mess with someone's project while sleepy, sorry_

SkyaTura commented:
So, I'm thinking and experimenting things for long hours now, but I can't find an answer.

What have I done since my first changes:

- In fact, this formatter had nothing to do with "middle attributes", so I've added a new option `pugSortAttributes = 'as-is' | 'asc' | 'desc'`. This could be reverted if you think it shouldn't exist.
- Added more test cases, with different middle attribute sorting options.

What I cannot understand is, why on earth `compare 3` and `compare 6` are failing.

Shinigami92 commented on src/options/attribute-sorting/index.ts:
```diff
@@ -18,3 +18,19 @@ export const PUG_SORT_ATTRIBUTES_END_OPTION = {
+	since: '1.8.0',
+	category: CATEGORY_PUG,
+	type: 'choice',
+	array: false,
```
You don't need to pass `array: false`, cause it's the default

Shinigami92 commented:
Can you rename the PR title? :)

Shinigami92 commented:
@SkyaTura Looking at the tests, it seems that your sort is not stable :thinking:

SkyaTura commented:
Yes, but I don't know why. My knowledge is limited here, because I though it should work correctly.

I'll study more about the sort method to understand what is going on, over this weekend, _( the last couple days have been busy for me :x )_

lehni commented:
Could it be the cases where the compare function returns `0` that leads to unstable sort?

Node's `sort()` wasn't stable up to about Node 11 or so.

Some bits about this:

https://bugs.chromium.org/p/v8/issues/detail?id=90
https://github.com/nodejs/node/issues/29446#issuecomment-528164143
https://medium.com/@fsufitch/is-javascript-array-sort-stable-46b90822543f

I would try a stable `sort()` method next, e.g.:

```js
function stableSort(
  array,
  compare = (a, b) => a < b ? -1 : a > b ? 1 : 0
 ) {
  const entries = array.map((value, index) => [value, index]);
  entries.sort((a, b) => {
    const order = compare(a[0], b[0]);
    // When order is 0, sort by index to make the sort stable:
    return order != 0 ? order : a[1] - b[1];
  });
  return entries.map(([value]) => value);
}
```

Note that this doesn't modify `array`, it returns a copy.

SkyaTura commented:
Finally!!! My logic wasn't wrong after all 😅 

Thank you a lot @lehni . This stableSort worked as a chamr.

Shinigami92 commented:
@SkyaTura @lehni Check out my newly added contributing.me 😃
Maybe it could help 🙂

lehni commented:
@SkyaTura happy to hear that this helped! I was struggling with this situation a while back on Node also, it can be so confusing.

Shinigami92 commented on src/options/attribute-sorting/utils.ts:
```diff
@@ -1,68 +1,77 @@
+	}
 
-		result = compareByIndex(aIndex, bIndex);
+	if (sortAttributes === 'asc') {
```
Can you convert this to a switch case? Then it could be extend better in future

Shinigami92 commented on src/options/attribute-sorting/utils.ts:
```diff
@@ -1,68 +1,77 @@
+	entries.sort((a, b) => {
+		const order = compare(a[0], b[0]);
+		// When order is 0, sort by index to make the sort stable:
+		return order != 0 ? order : a[1] - b[1];
```
Always prefer `!==`

Shinigami92 commented on src/options/attribute-sorting/utils.ts:
```diff
@@ -1,68 +1,77 @@
+	return 0;
+}
+
+export function stableSort<T>(array: T[], compare: CompareFunction): any[] {
```
Not quite sure, but isn't it `: T[]` as return type?

SkyaTura replied:
To be honest, I'm new to TypeScript, I didn't knew how to properly type the method @lehni suggested.

Could you help me with this one?

Shinigami92 replied:
Ah okay, I can checkout your branch at evening and then suggest some types

Shinigami92 commented on tests/options/sortAttributes/sort-attributes.test.ts:
```diff
@@ -38,19 +39,181 @@ describe('Options', () => {
+					pugSortAttributes,
+					pugSortAttributesBeginning,
+					pugSortAttributesEnd
+				)
```
Can you add a test using beginning and end but don't pass `pugSortAttributes` so it takes the default one?

Shinigami92 commented on tests/options/sortAttributes/sort-attributes.test.ts:
```diff
@@ -38,19 +39,181 @@ describe('Options', () => {
+
+			expect(actual).toStrictEqual(expected);
+		});
+		test('compare 5', () => {
```
I know it's hard, but I thing we should try to find names for all these `compare x` tests
Now there are 7 and we cannot read from them what they are doing ^^

Shinigami92 replied:
@SkyaTura Nice one, couldn't make it better ✨
